private/bufpkg/bufimage/bufimageutil/bufimageutil.go

```diff
@@ -1000,3 +1012,57 @@ func stripSourceRetentionOptionsFromFile(imageFile bufimage.ImageFile) (bufimage
 		imageFile.UnusedDependencyIndexes(),
 	)
 }
+
+// orderedImports is a structure to maintain an ordered set of imports. This is needed
+// because we want to be able to iterate through imports in a deterministic way when filtering
+// the image.
+type orderedImports struct {
+	pathToIndex map[string]int
+	paths       []string
+}
+
+// index returns the index for a given path. If the path does not exist in index map, -1
+// is returned and should be considered deleted.
+func (o *orderedImports) index(path string) int {
+	if index, ok := o.pathToIndex[path]; ok {
+		return index
+	}
+	return -1
+}
+
+// add appends a path to the paths list and the index in the map. If a key already exists,
+// then this is a no-op.
+func (o *orderedImports) add(path string) {
+	if _, ok := o.pathToIndex[path]; !ok {
+		o.paths = append(o.paths, path)
+		o.pathToIndex[path] = len(o.paths)
+	}
+}
+
+// delete removes a key from the index map of ordered imports. Note that the path is not removed
+// from the paths list. If you want to iterate through the paths, use keys() to get all
+// non-deleted keys.
+func (o *orderedImports) delete(path string) error {
+	if _, ok := o.pathToIndex[path]; ok {
+		delete(o.pathToIndex, path)
+		return nil
+	}
+	return fmt.Errorf("no import found for: %s", path)
+}
+
+// keys provides all non-deleted keys from the ordered imports.
+func (o *orderedImports) keys() []string {
+	keys := make([]string, 0, len(o.pathToIndex))
+	for _, path := range o.paths {
+		if _, ok := o.pathToIndex[path]; ok {
+			keys = append(keys, path)
+		}
+	}
+	return keys
```

Comment:
As we always build a new set of keys we could remove the storing the `paths` field in `orderedImports` and build it up with a sort, naming `pathToIndex` as `pathToWeight`, something like:
```go
paths := slices.Collect(maps.Keys(m))
sort.Slice(paths, func(i, j int) bool { return pathWeight[paths[i]] < pathWeight[paths[j]] })
```
Then the delete function doesn't need to be special cased.